### PR TITLE
Local teleporter deploy

### PR DESCRIPTION
**Pull request by felipemadero**
Closes https://github.com/ava-labs/avalanche-cli/issues/1469 https://github.com/ava-labs/avalanche-cli/issues/1380

### CODE REVIEW

File: cmd/networkcmd/start.go [hunk 1]
```diff
@@ -149,6 +153,32 @@ func StartNetwork(*cobra.Command, []string) error {
 		ux.PrintTableEndpoints(resp.ClusterInfo)
 	}
 
+	relayerStoredConfigPath := filepath.Join(app.GetAWMRelayerSnapshotConfsDir(), snapshotName+jsonExt)
```
Comment (arturrez) [inline cmd/networkcmd/start.go:156]: nit: assign `var snapshotFileName` `snapshotName+jsonExt` to make code more readable. in this case you don't need jsonExt constant as it's used only once as far as I can see

File: cmd/networkcmd/start.go [hunk 2]
```diff
@@ -149,6 +153,32 @@ func StartNetwork(*cobra.Command, []string) error {
...
+	if utils.FileExists(relayerStoredConfigPath) {
+		relayerConfigPath := app.GetAWMRelayerConfigPath()
+		if err := binutils.CopyFile(relayerStoredConfigPath, relayerConfigPath); err != nil {
+			return err
+		}
+		fmt.Println("")
```
Comment (arturrez) [inline cmd/networkcmd/start.go:162]: should we use `ux.Logger.PrintToUser` instead

File: cmd/networkcmd/stop.go [hunk 3]
```diff
@@ -42,6 +45,28 @@ func StopNetwork(*cobra.Command, []string) error {
 		return nil
 	}
 
+	relayerConfigPath := app.GetAWMRelayerConfigPath()
+	if utils.FileExists(relayerConfigPath) {
+		relayerStoredConfigPath := filepath.Join(app.GetAWMRelayerSnapshotConfsDir(), snapshotName+".json")
```
[the remark this answers is not in the document]
Comment (arturrez) [reply cmd/networkcmd/stop.go:50]: i feel like https://github.com/ava-labs/avalanche-cli/pull/1471/files#r1503271794 is a better approach for this. but it's just a nit anyway

File: cmd/subnetcmd/create.go [hunk 4]
```diff
@@ -218,6 +235,48 @@ func createSubnetConfig(cmd *cobra.Command, args []string) error {
 		return errors.New("not implemented")
 	}
 
+	if teleporterReady {
+		isSubnetEVMGenesis, err := jsonIsSubnetEVMGenesis(genesisBytes)
```
Comment (arturrez) [inline cmd/subnetcmd/create.go:239]: nit: replace with 
```
if isSubnetEVMGenesis, err := ...; err!=nil`; 
else if !isSubnetEVMGenesis
```
as `isSubnetEVMGenesis`  is not used anywhere else

File: cmd/subnetcmd/create.go [hunk 5]
```diff
@@ -218,6 +235,48 @@ func createSubnetConfig(cmd *cobra.Command, args []string) error {
...
+			if err := k.Save(keyPath); err != nil {
+				return err
+			}
+		}
+		ux.Logger.PrintToUser("  (evm address, genesis balance) = (%s, %v)", k.C(), teleporter.TeleporterPrefundedAddressBalance)
+		genesisBytes, err = addSubnetEVMGenesisPrefundedAddress(genesisBytes, k.C(), teleporter.TeleporterPrefundedAddressBalance.String())
+		if err != nil {
+			return err
+		}
+		// lets' use latest versions for teleporter contract and default awm relayer
```
Comment (arturrez) [inline cmd/subnetcmd/create.go:269]: nit: let's

File: cmd/subnetcmd/deploy.go [hunk 6]
```diff
@@ -381,7 +381,7 @@ func deploySubnet(cmd *cobra.Command, args []string) error {
 		}
 
 		deployer := subnet.NewLocalDeployer(app, userProvidedAvagoVersion, avagoBinaryPath, vmBin)
-		subnetID, blockchainID, err := deployer.DeployToLocalNetwork(chain, chainGenesis, genesisPath)
+		subnetID, blockchainID, teleporterMessengerAddress, teleporterRegistryAddress, err := deployer.DeployToLocalNetwork(chain, chainGenesis, genesisPath)
```
[the remark this answers is not in the document]
Comment (arturrez) [reply cmd/subnetcmd/deploy.go:384]: +1 for the structure to return

File: cmd/subnetcmd/deploy_test.go [hunk 7]
```diff
@@ -161,7 +161,7 @@ func TestCheckForInvalidDeployAndSetAvagoVersion(t *testing.T) {
 			networkErr:      nil,
 			desiredRPC:      19,
 			desiredVersion:  testLatestAvagoVersion,
-			expectError:     true,
+			expectError:     false,
```
Comment (arturrez) [inline cmd/subnetcmd/deploy_test.go:164]: is it a separate fix for e2e?  and it's not relevant to `teleporter` if I understand this code correctly

File: pkg/evm/evm.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+	})
+	txSigner := types.LatestSignerForChainID(chainID)
+	signedTx, err := types.SignTx(tx, txSigner, sourceAddressPrivateKey)
+	if err != nil {
+		return err
+	}
+	if err := client.SendTransaction(ctx, signedTx); err != nil {
+		return err
+	}
+	_, b, err := WaitForTransaction(client, signedTx)
```
Comment (arturrez) [inline pkg/evm/evm.go:119]: nit: can be reduced to
```
 if _, b, err := WaitForTransaction(client, signedTx); err!=nil
 else if !b
 else return nil
 ```

File: pkg/evm/evm.go [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+	tx := new(types.Transaction)
+	if err := tx.UnmarshalBinary(common.FromHex(txStr)); err != nil {
+		return err
+	}
+	ctx, cancel := utils.GetAPIContext()
+	defer cancel()
+	if err := client.SendTransaction(ctx, tx); err != nil {
+		return err
+	}
+	_, b, err := WaitForTransaction(client, tx)
```
Comment (arturrez) [inline pkg/evm/evm.go:142]: can `WaitForTransaction(client,tx) ` be replaced with `client.WaitForTransaction(tx)` to be consistent with ` client.SendTransaction`

File: pkg/evm/evm.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+	if err != nil {
```
Comment (arturrez) [inline pkg/evm/evm.go:143]: similar nit https://github.com/ava-labs/avalanche-cli/pull/1471/files#r1503314037

File: pkg/teleporter/relayer.go [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+		// so much expected after a reboot
+		if err := os.Remove(runFilePath); err != nil {
+			return fmt.Errorf("failed removing relayer run file %s: %w", runFilePath, err)
+		}
+		return nil
+	}
+	if err := proc.Signal(os.Interrupt); err != nil {
+		return fmt.Errorf("failed killing relayer process with pid %d: %w", rf.Pid, err)
+	}
+	if err := os.Remove(runFilePath); err != nil {
```
Comment (arturrez) [inline pkg/teleporter/relayer.go:151]: should we also try sending `Kill` signal to the process

File: pkg/teleporter/relayer.go [hunk 12]
```diff
@@ -0,0 +1,377 @@
...
+	}
+	proc, err := os.FindProcess(rf.Pid)
+	if err != nil {
+		// so much expected after a reboot
+		if err := os.Remove(runFilePath); err != nil {
+			return fmt.Errorf("failed removing relayer run file %s: %w", runFilePath, err)
+		}
+		return nil
+	}
+	if err := proc.Signal(syscall.Signal(0)); err != nil {
```
Comment (arturrez) [inline pkg/teleporter/relayer.go:141]: why do we send `syscall.Signal(0)`. is it a just a liveness check for the process to see if it exists. 
If so why we rm pid file and return? 
Or maybe I don't understand this code

File: pkg/teleporter/relayer.go [hunk 13]
```diff
@@ -0,0 +1,377 @@
...
+		return fmt.Errorf("failed removing relayer run file %s: %w", runFilePath, err)
+	}
+	return nil
+}
+
+func saveRelayerRunFile(runFilePath string, pid int) error {
+	rf := relayerRunFile{
+		Pid: pid,
+	}
+	bs, err := json.Marshal(&rf)
```
Comment (arturrez) [inline pkg/teleporter/relayer.go:161]: does `bs` stand for `bytes` ?

File: pkg/teleporter/relayer.go [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+	return nil
+}
+
+func installRelayer(binDir, version string) (string, error) {
+	binPath := filepath.Join(binDir, constants.AWMRelayerBin)
+	if utils.IsExecutable(binPath) {
+		return binPath, nil
+	}
+	ux.Logger.PrintToUser("Installing AWM-Relayer %s", version)
+	url, err := getRelayerURL(version)
```
Comment (arturrez) [inline pkg/teleporter/relayer.go:177]: should `getRelayerURL` return file ext so it can be passed instead of hardcoding `"tar.gz"`

File: pkg/teleporter/relayer.go [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+		BlockchainID:      blockchainID,
+		VM:                config.EVM.String(),
+		EncryptConnection: false,
+		APINodeHost:       host,
+		APINodePort:       port,
+		AccountPrivateKey: relayerFundedAddressKey,
+	}
+	sources := relayerConfig.SourceSubnets
+	found := false
+	for _, s := range sources {
```
Comment (arturrez) [inline pkg/teleporter/relayer.go:341]: nit: replace with utils.Filter and check len(filteredList)>0

File: pkg/teleporter/relayer.go [hunk 16]
```diff
@@ -0,0 +1,377 @@
...
+			found = true
+		}
+	}
+	if !found {
+		sources = append(sources, source)
+		relayerConfig.SourceSubnets = sources
+	}
+	destinations := relayerConfig.DestinationSubnets
+	found = false
+	for _, d := range destinations {
```
Comment (arturrez) [inline pkg/teleporter/relayer.go:352]: same as above

File: pkg/subnet/local.go [hunk 17]
```diff
@@ -972,3 +1100,24 @@ func CheckNodeIsInSubnetValidators(subnetID ids.ID, nodeID string) (bool, error)
...
+
+func GetChainIDs(endpoint string, chainName string) (string, string, error) {
+	pClient := platformvm.NewClient(endpoint)
+	ctx, cancel := utils.GetAPIContext()
+	defer cancel()
+	blockChains, err := pClient.GetBlockchains(ctx)
+	if err != nil {
+		return "", "", err
+	}
+	for _, chain := range blockChains {
```
Comment (arturrez) [inline pkg/subnet/local.go:1112]: nit: use `utils.Filter` for this

File: pkg/teleporter/relayer.go [hunk 18]
```diff
@@ -0,0 +1,377 @@
...
+	if !found {
+		destinations = append(destinations, destination)
+		relayerConfig.DestinationSubnets = destinations
+	}
+}
+
+// Get the host and port from a URI. The URI should be in the format http://host:port or https://host:port or host:port
+func getURIHostAndPort(uri string) (string, uint32, error) {
+	trimmedURI := uri
+	trimmedURI = strings.TrimPrefix(trimmedURI, "http://")
```
Comment (arturrez) [inline pkg/teleporter/relayer.go:366]: use `url.Parse` instead

File: pkg/teleporter/teleporter.go [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+	prefundedPrivateKey string,
+) (bool, string, string, error) {
+	alreadyDeployed, messengerAddress, err := t.DeployMessenger(teleporterInstallDir, version, subnetName, rpcURL, prefundedPrivateKey)
+	if err != nil {
+		return false, "", "", err
+	}
+	if alreadyDeployed {
+		return true, messengerAddress, "", nil
+	}
+	registryAddress, err := t.DeployRegistry(teleporterInstallDir, version, subnetName, rpcURL, prefundedPrivateKey)
```
Comment (arturrez) [inline pkg/teleporter/teleporter.go:137]: can be reduced to
``` if err:=; err!=nil
else return```

File: pkg/teleporter/teleporter.go [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+) (bool, string, error) {
+	if err := t.DownloadAssets(teleporterInstallDir, version); err != nil {
+		return false, "", err
+	}
+	// check if contract is already deployed
+	client, err := evm.GetClient(rpcURL)
+	if err != nil {
+		return false, "", err
+	}
+	teleporterMessengerAlreadyDeployed, err := evm.ContractAlreadyDeployed(client, t.teleporterMessengerContractAddress)
```
Comment (arturrez) [inline pkg/teleporter/teleporter.go:159]: as `teleporterMessengerAlreadyDeployed` is not used anywhere else if can be reduced to 
```
if teleporterMessengerAlreadyDeployed, err :=; err!=nil
bla, return
else if teleporterMessengerAlreadyDeployed bla, return bla
```

File: pkg/teleporter/teleporter.go [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+	defer client.Close()
+	signer, err := evm.GetSigner(client, prefundedPrivateKey)
+	if err != nil {
+		return "", err
+	}
+	teleporterRegistryAddress, tx, _, err := teleporterRegistry.DeployTeleporterRegistry(signer, client, teleporterRegistryConstructorInput)
+	if err != nil {
+		return "", err
+	}
+	_, success, err := evm.WaitForTransaction(client, tx)
```
Comment (arturrez) [inline pkg/teleporter/teleporter.go:221]: nit: as it's success is not used; it can be reduced similar to above

File: tests/e2e/commands/subnet.go [hunk 22]
```diff
@@ -38,7 +38,16 @@ func CreateSubnetEvmConfigWithVersion(subnetName string, genesisPath string, ver
...
-	cmdArgs := []string{SubnetCmd, "create", "--genesis", genesisPath, "--evm", subnetName, "--" + constants.SkipUpdateFlag}
+	cmdArgs := []string{
+		SubnetCmd,
+		"create",
+		"--genesis",
+		genesisPath,
+		"--evm",
+		subnetName,
+		"--" + constants.SkipUpdateFlag,
+		"--teleporter=false",
```
Comment (arturrez) [inline tests/e2e/commands/subnet.go:49]: make teleporter=true|false param to pass to it can be used in e2e

File: tests/e2e/commands/subnet.go [hunk 23]
```diff
@@ -118,6 +127,7 @@ func CreateCustomVMConfig(subnetName string, genesisPath string, vmPath string)
 		"--custom-vm-path",
 		vmPath,
 		"--"+constants.SkipUpdateFlag,
+		"--teleporter=false",
```
Comment (arturrez) [inline tests/e2e/commands/subnet.go:130]: same

File: tests/e2e/testcases/subnet/local/suite.go [hunk 24]
```diff
@@ -242,14 +242,17 @@ var _ = ginkgo.Describe("[Local Subnet]", ginkgo.Ordered, func() {
...
 		gomega.Expect(err).Should(gomega.BeNil())
 		gomega.Expect(rpcs).Should(gomega.HaveLen(1))
 
-		deployOutput = commands.DeploySubnetLocally(subnetName)
+		out, err := commands.DeploySubnetLocallyWithArgsAndOutput(subnetName, "", "")
+		gomega.Expect(err).Should(gomega.HaveOccurred())
+		deployOutput = string(out)
 		rpcs, err = utils.ParseRPCsFromOutput(deployOutput)
 		if err == nil {
 			fmt.Println(deployOutput)
```
Comment (arturrez) [inline tests/e2e/testcases/subnet/local/suite.go:258]: we need e2e for teleporter

File: cmd/subnetcmd/create.go [hunk 25]
```diff
@@ -218,6 +235,48 @@ func createSubnetConfig(cmd *cobra.Command, args []string) error {
...
+		teleporterVersion, err := app.Downloader.GetLatestReleaseVersion(binutils.GetGithubLatestReleaseURL(constants.AvaLabsOrg, constants.TeleporterRepoName))
+		if err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/subnetcmd/create.go:273]: There is no compatiblity file. I believe may not be needed. It is mainly evm code and related to the underlying warp precompile. Surely there is a dependency between warp implementation and teleporter implementation. Not sure if warp implementation is completelly fixed? cc @michaelkaplan13

[the remark this answers is not in the document]
Comment (michaelkaplan13) [reply cmd/subnetcmd/create.go:273]: Yep, that is correct. Teleporter only depends on the warp precompile being available.

I'm hesitant to say anything is completely fixed, but the warp precompile can only be changed with a required network upgrade, and any non-backwards compatible changes would potentially break any contracts that use it (including Teleporter), so that would be avoid at the EVM level at basically all costs.

TL;DR: I don't think you need to worry about version compatibility here as long as the blockchain has the warp precompile enabled. 👍

File: pkg/subnet/local.go [hunk 26]
```diff
@@ -540,10 +550,128 @@ func (d *LocalDeployer) doDeploy(chain string, chainGenesis []byte, genesisPath
...
+				CChainTeleporterMessengerAddress: cchainTeleporterMessengerAddress,
+				CChainTTeleporterRegistryAddress: cchainTeleporterRegistryAddress,
+			}
+			bs, err := json.Marshal(&extraLocalNetworkData)
+			if err != nil {
+				return ids.Empty, ids.Empty, "", "", err
+			}
+			if err := os.WriteFile(extraLocalNetworkDataPath, bs, constants.WriteReadReadPerms); err != nil {
+				return ids.Empty, ids.Empty, "", "", fmt.Errorf("could not extra local network data file to %s: %w", extraLocalNetworkDataPath, err)
+			}
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply pkg/subnet/local.go:626]: let's just keep it as is

File: pkg/subnet/local.go [hunk 27]
```diff
@@ -540,10 +550,128 @@ func (d *LocalDeployer) doDeploy(chain string, chainGenesis []byte, genesisPath
...
+			teleporterRegistryAddress,
+		); err != nil {
+			return ids.Empty, ids.Empty, "", "", err
+		}
+		fmt.Println()
+		// fund relayer on current blockchain
+		if err := teleporter.FundRelayer(endpointRPCURL, privKeyStr, relayerAddress); err != nil {
+			return ids.Empty, ids.Empty, "", "", err
+		}
+		// start relayer
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply pkg/subnet/local.go:664]: same here

File: cmd/networkcmd/start.go (hunk 1 above)
Comment (felipemadero) [reply cmd/networkcmd/start.go:156]: will just keep it as is

File: cmd/subnetcmd/deploy_test.go (hunk 7 above)
Comment (felipemadero) [reply cmd/subnetcmd/deploy_test.go:164]: this is a fix for unit testing. while testing teleporter and upgrading to durango issues were found on compatibility checks, so a change was made to CheckForInvalidDeployAndGetAvagoVersion:
- now removes the rc part of the provided avago version for comparisons
- if no avago version was found for a given subnet, it returns pre release (which should be most probablty situation of avago prerelease using a new rpc version not found in compatibility file)

File: pkg/teleporter/relayer.go [hunk 28]
```diff
@@ -0,0 +1,377 @@
...
+	if err != nil {
+		return err
+	}
+	rf := relayerRunFile{}
+	if err := json.Unmarshal(bs, &rf); err != nil {
+		return err
+	}
+	proc, err := os.FindProcess(rf.Pid)
+	if err != nil {
+		// so much expected after a reboot
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply pkg/teleporter/relayer.go:135]: added a comment

File: pkg/teleporter/relayer.go (hunk 12 above)
Comment (felipemadero) [reply pkg/teleporter/relayer.go:141]: it is a check to see if the process exists. FindProcess is not enough

File: pkg/teleporter/relayer.go (hunk 13 above)
Comment (felipemadero) [reply pkg/teleporter/relayer.go:161]: yes

File: pkg/teleporter/relayer.go (hunk 14 above)
Comment (felipemadero) [reply pkg/teleporter/relayer.go:177]: not needed, always tar.gz

File: pkg/teleporter/relayer.go (hunk 15 above)
Comment (felipemadero) [reply pkg/teleporter/relayer.go:341]: do not agree to create a list to check if a element satisfy a condition, probably we can add a generic function to search for a condition in a list

File: pkg/subnet/local.go (hunk 17 above)
Comment (felipemadero) [reply pkg/subnet/local.go:1112]: I should prefer to use a utils.Find

File: tests/e2e/commands/subnet.go (hunk 22 above)
Comment (felipemadero) [reply tests/e2e/commands/subnet.go:49]: will do for teleporter e2e test

File: tests/e2e/commands/subnet.go (hunk 23 above)
Comment (felipemadero) [reply tests/e2e/commands/subnet.go:130]: same

File: tests/e2e/testcases/subnet/local/suite.go (hunk 24 above)
Comment (felipemadero) [reply tests/e2e/testcases/subnet/local/suite.go:258]: will add it together with teleporter msg command
